// @dainprotocol/service-sdk/dist/service/index.js
"use strict";
// File: src/service/index.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoreUtils = exports.createToolbox = exports.createService = exports.createTool = exports.defineCloudflareService = exports.defineDenoService = exports.defineNodeService = exports.defineDAINService = void 0;
const tslib_1 = require("tslib");
const nodeService_1 = require("./nodeService");
Object.defineProperty(exports, "defineNodeService", { enumerable: true, get: function () { return nodeService_1.defineDAINService; } });
const denoService_1 = require("./denoService");
Object.defineProperty(exports, "defineDenoService", { enumerable: true, get: function () { return denoService_1.defineDAINService; } });
const cloudflareService_1 = require("./cloudflareService");
Object.defineProperty(exports, "defineCloudflareService", { enumerable: true, get: function () { return cloudflareService_1.defineDAINService; } });
const core_1 = require("./core");
Object.defineProperty(exports, "createTool", { enumerable: true, get: function () { return core_1.createTool; } });
Object.defineProperty(exports, "createService", { enumerable: true, get: function () { return core_1.createService; } });
Object.defineProperty(exports, "createToolbox", { enumerable: true, get: function () { return core_1.createToolbox; } });
Object.defineProperty(exports, "CoreUtils", { enumerable: true, get: function () { return core_1.CoreUtils; } });
const defineDAINService = (config) => {
    throw new Error("This is a fallback implementation. Use the appropriate runtime-specific import.");
};
exports.defineDAINService = defineDAINService;
// Re-export all types
tslib_1.__exportStar(require("./types"), exports);
