// @dainprotocol/service-sdk/dist/service/nodeService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoreUtils = exports.createToolbox = exports.createService = exports.createTool = void 0;
exports.defineDAINService = defineDAINService;
const tslib_1 = require("tslib");
const service_1 = require("./service");
const core_1 = require("./core");
Object.defineProperty(exports, "createTool", { enumerable: true, get: function () { return core_1.createTool; } });
Object.defineProperty(exports, "createService", { enumerable: true, get: function () { return core_1.createService; } });
Object.defineProperty(exports, "createToolbox", { enumerable: true, get: function () { return core_1.createToolbox; } });
Object.defineProperty(exports, "CoreUtils", { enumerable: true, get: function () { return core_1.CoreUtils; } });
function defineDAINService(config) {
    const baseService = (0, service_1.createBaseDAINService)(config);
    const startNode = async (options = {
        port: 3000,
    }) => {
        const { port } = options;
        const { serve } = await Promise.resolve().then(() => tslib_1.__importStar(require("@hono/node-server")));
        const service = baseService.startService();
        const server = serve({
            fetch: service.fetch,
            port
        });
        console.log(`Server is listening on port ${port} -- Node Environment`);
        console.log(`DAIN Service is running on version ${config.metadata.version}`);
        return {
            address: () => ({ port }),
            shutdown: async () => {
                console.log("Shutting down DAIN service...");
                server.close();
            },
        };
    };
    return {
        ...baseService,
        startNode,
    };
}
// Re-export all types
tslib_1.__exportStar(require("./types"), exports);
