// @dainprotocol/service-sdk/dist/client/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServicePinnableWithWidgetSchema = exports.ServicePinnableSchema = exports.GetAllToolsAsJsonSchemaResponseSchema = exports.ToolWithJsonSchemaSchema = exports.ServiceContextWithDataSchema = exports.ServiceContextSchema = exports.ToolConfigSchema = exports.ToolboxConfigSchema = exports.ServiceConfigSchema = exports.MetadataSchema = exports.PricingInfoSchema = void 0;
const zod_1 = require("zod");
// Define and export types for DAIN service responses
exports.PricingInfoSchema = zod_1.z.object({
    pricePerUse: zod_1.z.number(),
    currency: zod_1.z.string(),
});
exports.MetadataSchema = zod_1.z.object({
    title: zod_1.z.string(),
    description: zod_1.z.string(),
    version: zod_1.z.string(),
    logo: zod_1.z.string().optional(),
    author: zod_1.z.string(),
    tags: zod_1.z.array(zod_1.z.string()),
});
exports.ServiceConfigSchema = zod_1.z.object({
    id: zod_1.z.string(),
    name: zod_1.z.string(),
    description: zod_1.z.string(),
    metadata: zod_1.z.object({
        capabilities: zod_1.z.array(zod_1.z.string()),
        languages: zod_1.z.array(zod_1.z.string()),
    }),
    recommendedPrompt: zod_1.z.string(),
    recommendedTools: zod_1.z.array(zod_1.z.string()),
});
exports.ToolboxConfigSchema = zod_1.z.object({
    id: zod_1.z.string(),
    name: zod_1.z.string(),
    description: zod_1.z.string(),
    tools: zod_1.z.array(zod_1.z.string()),
    metadata: zod_1.z.object({
        complexity: zod_1.z.string(),
        applicableFields: zod_1.z.array(zod_1.z.string()),
    }),
    recommendedPrompt: zod_1.z.string(),
});
exports.ToolConfigSchema = zod_1.z.object({
    id: zod_1.z.string(),
    name: zod_1.z.string(),
    description: zod_1.z.string(),
    pricing: exports.PricingInfoSchema,
    inputSchema: zod_1.z.any(), // This will be parsed separately
    outputSchema: zod_1.z.any(), // This will be parsed separately
});
exports.ServiceContextSchema = zod_1.z.object({
    id: zod_1.z.string(),
    name: zod_1.z.string(),
    description: zod_1.z.string(),
});
exports.ServiceContextWithDataSchema = exports.ServiceContextSchema.extend({
    data: zod_1.z.any(),
});
exports.ToolWithJsonSchemaSchema = zod_1.z.object({
    id: zod_1.z.string(),
    name: zod_1.z.string(),
    description: zod_1.z.string(),
    inputSchema: zod_1.z.any(),
});
exports.GetAllToolsAsJsonSchemaResponseSchema = zod_1.z.object({
    tools: zod_1.z.array(exports.ToolWithJsonSchemaSchema),
    reccomendedPrompts: zod_1.z.array(zod_1.z.string()),
});
exports.ServicePinnableSchema = zod_1.z.object({
    id: zod_1.z.string(),
    type: zod_1.z.enum(["button", "widget"]),
    label: zod_1.z.string(),
    name: zod_1.z.string(),
    description: zod_1.z.string(),
    icon: zod_1.z.string(),
});
exports.ServicePinnableWithWidgetSchema = exports.ServicePinnableSchema.extend({
    widget: zod_1.z.object({
        text: zod_1.z.string().optional(),
        data: zod_1.z.any().optional(),
        ui: zod_1.z.any().optional(),
    }),
});
